### ISSUE TITLE
Question on status and missing documentation

### ISSUE THREAD

**Issue by wez** [adamreichold/b2_backup#1]
Hi! I found this while searching for incremental backup solutions for b2.

It seems like you've been working on this for a while, but it's not clear if this is something you personally rely on or if it is potentially a bit shaky, and it's light on docs.

What is the status of the project, and could you give a brief summary of how to use it?

Thanks!

**Comment by adamreichold** [comment]
> It seems like you've been working on this for a while, but it's not clear if this is something you personally rely on or if it is potentially a bit shaky, and it's light on docs.

I use this program personally to backup some smaller but important-to-me datasets. The original motivation was to have something which is overall simple, works well against the limited API provided by B2 (e.g. does not need to read back any data during incremental backup) and where I would be sure that the encryption happens locally.

The lack of documentation is somewhat intentional as while I rely on this personally, I am not sure whether I will be able to maintain this if other people start to use it more widely. It also takes some shortcuts which I would probably need to resolve for wider adoption, e.g. the file format is currently not versioned and I just did full backups when I made any changes in the past. (I think that adding the adding the metadata for version and checking it would be simple enough, but committing to actually continue to support old versions is something else entirely.)

> What is the status of the project, and could you give a brief summary of how to use it?

So currently, it is a tool I personally rely on and make available without any commitments w.r.t. maintenance.

As for usage, I usually put the binary into a folder `$HOME/bin/b2_backup` where also the database `manifest` and the configuration `config.toml` will be stored. I use the timer and service unit provided in the `systemd` folder of this repo, put them into `$HOME/.config/systemd/user` and enable them via `systemctl --user enable --now b2_backup.timer` to run this as a daily user service.

(Or basically setup in the same way with the data in `/var/lib/b2_backup` and as a system service for server usage.)

(You will probably need to install `schedtool` if you want to use these units as they are. But that invocation can also be removed from them without loss of function.)

A minimal configuration file would look like the following

```yaml
app_key_id: '0000000000000000000000000'
app_key: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'
bucket_id: 'aaaaaaaaaaaaaaaaaaaaaaaa'
bucket_name: 'foo'
key: 'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb'
includes:
  - /home/bar/Documents
  - /home/bar/.thunderbird
excludes: []
```

and use 0600 permissions. The application key and bucket need to be manually created via the B2 UI or API. `includes` lists directories which will be recursively added to the backup while `excludes` lists subdirectories of those which will not be recursed into.

One thing that would probably need additional automation is garbage collection, i.e. I currently manually run `b2_backup collect-small-archives` and `b2_backup collect-small-patchsets` at irregular intervals. As their names suggest, they garbage collect archives/patchsets which contain mostly stale data and merge them into archives/patchsets which do not. To do that, they need to download from B2 which might incur additional fees and hence they should be run seldom, e.g. once a month would probably suffice which is also why I did not bother writing automation yet.

**Comment by wez** [comment]
Thanks for the overview!
I totally understand your position regarding support!

I haven't gotten hands-on with this yet, but if/when I do, it's possible that I might send a PR or two your way, eg: perhaps adding glob style matching for includes/excludes, and perhaps also supporting gitignore style exclusion.  I'm also considering using zfs properties to label filesystems for inclusion/exclusion: I think I can have an external tool generate a config fragment for that and then have `b2_backup` accept that in addition to the main config, rather than teaching it directly about zfs.  Needs a bit more thought, and again: I haven't really sat down to look at this in earnest yet, and I may still go in a different direction!

Would you be open to accepting those sorts of PRs?

**Comment by adamreichold** [comment]
> Would you be open to accepting those sorts of PRs?

Adding support for globs and ignores via the well-known crates for these tasks seem like reasonable additions to me and I would be glad to review such PR.

As written above, I do value simplicity which does include keeping the dependency closure manageable but this does not preclude adding features, especially if they are well localised like what you describe.

**Comment by adamreichold** [comment]
> I'm also considering using zfs properties to label filesystems for inclusion/exclusion: I think I can have an external tool generate a config fragment for that and then have b2_backup accept that in addition to the main config, rather than teaching it directly about zfs.

I think a schema like this could be very flexible, but I would like to avoid being tied to a specific file system. Are you dependent on using ZFS properties or would a mechanism based on the presence of a custom extended attribute which we could check for using `getxattr` be sufficient as well?

**Comment by wez** [comment]
As a concept, I think extended attributes would work and be useful, but there is an important difference between those and zfs filesystem properties: with zfs, it is possible to enumerate all filesystems and their properties without traversing the entire filesystem.  zfs filesystems are hierarchical, and can be easily carved out of the storage pool and configured with different properties in different locations.  So I could have `/foo` and `/foo/bar/must-backup` as two filesystems and discover them in "constant time" and without doing any significant I/O.

If I didn't want `/foo` to be backed up, but needed `/foo/bar/must-backup` it would be very cheap to understand that from the zfs properties.

If that same scheme were encoded with extended attributes in one big filesystem (because it's not quite as easy to split up non-zfs storage in the same way as above), the include nested under an exclude wouldn't be discoverable without first walking all of `/foo`, but since it is marked as exclude it seems desirable to avoid traversing foo once the exclusion property is discovered.

Now, I don't know if I actively need to do exclude-but-include-something-nested-deeper; I think most of my stuff is include-but-exclude-something-nested-deeper, but I wanted to highlight that as a difference between the two schemes.

It might be desirable to provide the user some control/constraints for extended attribute based discovery, so that walking the entire VFS can be avoided.  And maybe a separate option that will enable the exclude-but-include pattern, but turn it off by default to save I/O?

Another thing that I think would be desirable, for peace of mind, is a way to quickly summarize perhaps from the manifest/db, the locations which were found to have the extended attributes attached and display some summary statistics from the most recent backup and/or aggregate over all backups (eg: total size and/or number of files under that location).

That way it would be possible to see at a glance which locations have been backed up and gut-check their sizes without having to run a `find` or something like that.  Having instant insight into what should get backed up feels important.  Right now you can simply look in the config file, but once you make it dynamic in some way you lose that unless you also add some way to review it.

So in summary: I think extended attributes support for this sort of thing is good, but I think I would prefer to annotate in zfs for my use case, and I don't mind having my own separate script that synthesizes an appropriate config based on its interrogation of zfs.

**Comment by adamreichold** [comment]
> So in summary: I think extended attributes support for this sort of thing is good, but I think I would prefer to annotate in zfs for my use case, and I don't mind having my own separate script that synthesizes an appropriate config based on its interrogation of zfs.

I agree with the performance and usability pitfalls of tagging individual directories/files using extend attributes and would say we should leave this on the table for now.

I did wonder whether shelling out from b2_backup to dynamically produce includes/excludes might be desirable, but I think the process you describe above is preferable due to the ability to introspect the resulting configuration.
